src/classes/ADDR_ProcessControl.cls

```diff
@@ -0,0 +1,100 @@
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+ * @author Salesforce.org
+ * @date 2017
+ * @group Address Management
+ * @description Process control class to prevent trigger recursion in async process for address handling
+ *
+***/
+public class ADDR_ProcessControl {
+
+    /*******************************************************************************************************
+    * @description Flag to true to prevent the AfterUpdate trigger on the Contact from processing changes
+    * to the Contact's address that are being made by this BeforeUpdate trigger. Only used by ADDR_Contact_TDTM.
+    */
+    public static Boolean hasAlreadyRunBeforeOnContact = false;
+
+    /*******************************************************************************************************
+    * @description Flag to prevent recursion into the Address trigger handler/logic
+    */
+    private static Integer recursionFlag = 0;
+
+    /*******************************************************************************************************
+    * @description Increment/Decrement the static flag counter that tracks whether the Address processing
+    *    trigger has been called already
+    * @param b A boolean where true increments the counter and false decrements it.
+    * @return void This method modifies class variables and doesn't return anything.
+    ********************************************************************************************************/
+    public static void setRecursionFlag(boolean b) {
+        recursionFlag += (b ? 1 : -1);
+        if (recursionFlag < 0) {
+            recursionFlag = 0;
+        }
+        UTIL_Debug.debug('>>>>>> setRecursionFlag(' + b + ') ==> ' + recursionFlag);
+    }
+
+    /* @description Returns true if the recursionFlag value is greater than 1 */
+    public static Boolean checkRecursionState() {
+        UTIL_Debug.debug('>>>>>> checkRecursionState() ==> ' + recursionFlag);
+        return (recursionFlag > 0);
+    }
+
+    /*******************************************************************************************************
+    * @description Simple static flag to prevent execution of the address validation triggers
```

Comment:
could you update the comment to be more clear.  this is about calling the address validation services, not validation within a trigger!  had me stumped for a bit why you'd add this flag as well as the separate recursionFlag.